dev/AppLifecycle/Association.cpp

```diff
@@ -0,0 +1,423 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+#include <pch.h>
+#include "Association.h"
+
+namespace winrt::Microsoft::ProjectReunion::implementation
+{
+    HKEY GetRegistrationRoot()
+    {
+        return HKEY_CURRENT_USER;
+    }
+
+    bool IsFileExtension(const std::wstring& extension)
+    {
+        return (extension.at(0) == L'.');
+    }
+
+    std::wstring GetFullIdentityString()
+    {
+        std::wstring identityString;
+        WCHAR idNameBuffer[PACKAGE_FULL_NAME_MAX_LENGTH+1];
+        UINT32 idNameBufferLen = ARRAYSIZE(idNameBuffer);
+        if (::GetCurrentPackageFullName(&idNameBufferLen, idNameBuffer) == ERROR_SUCCESS)
+        {
+            identityString = idNameBuffer;
+        }
+
+        return identityString;
+    }
+
+    bool HasIdentity()
+    {
+        return !(GetFullIdentityString()).empty();
+    }
+
+    std::wstring GetModulePath()
+    {
+        std::wstring path(100, L'?');
+        uint32_t path_size{};
+        DWORD actual_size{};
+
+        do
+        {
+            path_size = static_cast<uint32_t>(path.size());
+            actual_size = ::GetModuleFileName(nullptr, path.data(), path_size);
+
+            if (actual_size + 1 > path_size)
+            {
+                path.resize(path_size * 2, L'?');
+            }
+        } while (actual_size + 1 > path_size);
+
+        path.resize(actual_size);
+        return path;
+    }
+
+    std::wstring ComputeAppId()
+    {
+        // PRF = Project Reunion Framework.  A simple prefix to help identify ProgIds we compute here.
+        // AppId = Prefix + Hash(modulePath)
+        auto modulePath = GetModulePath();
+        std::hash<std::wstring> hasher;
+        auto hash = hasher(modulePath);
+;
+        wchar_t hashString[17]{}; // 16 + 1 characters for 64bit value represented as a string with a null terminator.
+        THROW_IF_FAILED(StringCchPrintf(hashString, sizeof(hashString), L"%I64x", hash));
```

Comment:
StringCchPrintf is "char count" based, sizeof is byte based.  Is this combo correct?

Reply:
Doh, definitely not correct.  Good catch!